Plain cmake

jontje opened the pull request:
Draft PR that turns the library into a plain cmake package, and it aims to solve issue https://github.com/ros-industrial/abb_libegm/issues/3.

The main reason for this is to be able to build and use the library regardless if e.g.:

- Ubuntu or Windows
- ROS or ROS2

traversaro commented on CMakeLists.txt:
```diff
@@ -1,52 +1,55 @@
+set(${PROJECT_NAME}_MAJOR_VERSION 1)
+set(${PROJECT_NAME}_MINOR_VERSION 0)
+set(${PROJECT_NAME}_PATCH_VERSION 0)
+set(${PROJECT_NAME}_VERSION ${${PROJECT_NAME}_MAJOR_VERSION}.${${PROJECT_NAME}_MINOR_VERSION}.${${PROJECT_NAME}_PATCH_VERSION})
```
As you are requiring using CMake 3.6, you can directly pass the `VERSION`  argument to the `project` , and this will automatically set the variable `${PROJECT_NAME}_VERSION`, see https://cmake.org/cmake/help/v3.6/command/project.html .

gavanderhoorn replied:
I would actually suggest to parse the version from the `package.xml`.

On platforms that don't care/ignore `package.xml`, it will not matter: they'll just see a normal version nr. On ROS/Catkin/Colcon based platforms, the manifest will be leading, which is what we'd want there.

traversaro replied:
Related code for parsing the version from package.xml : https://github.com/dartsim/dart/pull/1363/files .

gavanderhoorn replied:
I've done it like this in other projects:

```cmake
file(READ ${CMAKE_CURRENT_SOURCE_DIR}/package.xml package_xml_str)
if(NOT package_xml_str MATCHES "<version>([0-9]+.[0-9]+.[0-9]+)</version>")
  message(FATAL_ERROR "Could not parse project version from package.xml: cannot continue")
else()
  set(${PROJECT_NAME}_VERSION ${CMAKE_MATCH_1})
endif()
```

jontje replied:
Should major, minor and patch version still be set? Or is it enough with the `_VERSION ` variable?

gavanderhoorn replied:
Another alternative:

```cmake
# Read version from the package.xml file
file(READ ${CMAKE_CURRENT_SOURCE_DIR}/package.xml package_xml_str)
if(NOT package_xml_str MATCHES "<version>([0-9]+.[0-9]+.[0-9]+)</version>")
  message(FATAL_ERROR "Could not parse project version from package.xml: cannot continue")
endif()

# allow use of VERSION in project(..)
cmake_policy(SET CMP0048 NEW)

# at this point we either have a proper version string, or we've errored
# out with a FATAL_ERROR above. So assume CMAKE_MATCH_1 contains our
# package's version
project(abb_libegm VERSION ${CMAKE_MATCH_1} LANGUAGES CXX)
```

> Should major, minor and patch version still be set? Or is it enough with the _VERSION variable?

with the approach in this comment everything is set by/in `project(..)`, so no need to touch any of those variables yourself.

traversaro replied:
If `cmake_minimum_required(VERSION 3.5)` is used, I do not think that `cmake_policy(SET CMP0048 NEW)`  is needed, as it will be implicitly already set, see  https://cmake.org/cmake/help/latest/command/cmake_minimum_required.html#policy-settings .

gavanderhoorn replied:
The documentation on `project(..)` thoroughly confused me, as it made it seems like that policy was not set by default (I've used `VERSION` for as long as I can remember, so I was rather surprised to read about `CMP0048` here).

It's not needed.

traversaro commented on CMakeLists.txt:
```diff
@@ -55,30 +58,86 @@ set(SRC_FILES
-                    ${Boost_INCLUDE_DIRS}
-                    ${PROTOBUF_INCLUDE_DIRS}
-                    ${catkin_INCLUDE_DIRS})
+add_library(${PROJECT_NAME} SHARED ${SRC_FILES})
```
There is any specific reason why you won't leave the user the possibility of selecting the type of the library (`SHARED` or `STATIC`) using the standard `BUILD_SHARED_LIBS` variable? If you want to default to use `SHARED`, you can simply declare the `BUILD_SHARED_LIBS` variable as an option whose default value is `ON`, as suggested in CMake docs itself: https://cmake.org/cmake/help/v3.6/variable/BUILD_SHARED_LIBS.html . This would simplify packaging this lib for vcpkg.

jontje replied:
No reason, it was like that in the generated cmake package template (from the ROS2 command line interface) that I used as a base. I.e. the template found [here](https://github.com/ros2/ros2cli/tree/master/ros2pkg/ros2pkg/resource/cmake).

I like your arguments for making it an option instead.

traversaro replied:
> No reason, it was like that in the generated cmake package template (from the ROS2 command line interface) that I used as a base. I.e. the template found [here](https://github.com/ros2/ros2cli/tree/master/ros2pkg/ros2pkg/resource/cmake).
> 
> I like your arguments for making it an option instead.

Thanks, I know where to open an issue to ask to make it an option then. :D

traversaro commented on CMakeLists.txt:
```diff
@@ -55,30 +58,86 @@ set(SRC_FILES
+target_include_directories(${PROJECT_NAME} PUBLIC
+  "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include;${CMAKE_CURRENT_BINARY_DIR}>"
+  $<INSTALL_INTERFACE:include>
+  PUBLIC ${Boost_INCLUDE_DIRS}
```
Are you sure that this are necessary? In theory if you use the imported targets for both Boost and Protobuf, you can avoid setting explicitly the include dirs as "public", and make sure that your CMake package is relocatable, that is an extremly convenient property especially on Windows, for example if the library is included as part of an installer. See https://cmake.org/cmake/help/v3.6/manual/cmake-packages.7.html#creating-relocatable-packages for more info.

traversaro replied:
Sorry, the previous comment was not really clear. If it is feasible to bump the minimimum required version of CMake to 3.9, to link Boost and Protobuf you can use the imported targets `Boost::boost`, `Boost::system`, `Boost::thread`, `Boost::regex` and `protobuf::libprotobuf` to avoid to manually deal with include directories, and to have a relocatable CMake package. See https://cmake.org/cmake/help/v3.9/module/FindProtobuf.html and https://cmake.org/cmake/help/v3.9/module/FindBoost.html for more info.

gavanderhoorn replied:
Seeing as we're still targetting Xenial here, speccing `3.9` is not possible right now.

Which means we're stuck with using plain old CMake variables.

traversaro commented on CMakeLists.txt:
```diff
@@ -55,30 +58,86 @@ set(SRC_FILES
+  PUBLIC ${Boost_INCLUDE_DIRS}
+  PUBLIC ${Protobuf_INCLUDE_DIRS}
+)
+target_link_libraries(${PROJECT_NAME}
```
If Boost and Protobuf headers are used in the public headers of the libraries, I suggest to mark them as `PUBLIC` .

jontje replied:
Sorry for my lack of knowledge, but could you explain how to mark them as `PUBLIC`? Doesn't lines [68-69](https://github.com/jontje/abb_libegm/blob/plain_cmake/CMakeLists.txt#L68-L69) already do that?

traversaro replied:
> Sorry for my lack of knowledge, but could you explain how to mark them as `PUBLIC`? Doesn't lines [68-69](https://github.com/jontje/abb_libegm/blob/plain_cmake/CMakeLists.txt#L68-L69) already do that?

Lines [68-69](https://github.com/jontje/abb_libegm/blob/plain_cmake/CMakeLists.txt#L68-L69) tells the **preprocessor** of the downstream compilation unit that uses `abb_libegm` the **include directories** of Boost and Protobuf need to be appended to the search path of includes (tipically on GCC this is done via the `-I` compilation option).

Marking as PUBLIC (by doing `target_link_libraries(${PROJECT_NAME} PUBLIC ${Boost_LIBRARIES} ${Protobuf_LIBRARIES}`) the linked libraries, tells the **linker** of the downstream library that uses `abb_libegm`, to also link the Protobuf and Boost libraries when `abb_libegm` is linked. This is typically done (in *nix) via the `-l` linker option.
With CMake >= 3.9, you can just declare as `PUBLIC` in `target_link_libraries` the linking of the Boost's and Protobuf's imported targets, and that includes both passing the right library to the linker, and the right include dirs to the preprocessor, thanks to the use of so-called "Usage Requirements", see https://cmake.org/cmake/help/latest/manual/cmake-buildsystem.7.html#build-specification-and-usage-requirements . If you want a more in-depth discussion, the best reference for this that I ever found is the book https://crascit.com/professional-cmake/ , but it is not available free of charge.

traversaro commented on CMakeLists.txt:
```diff
@@ -55,30 +58,86 @@ set(SRC_FILES
+)
+
+# Create the ${PROJECT_NAME}Config.cmake.
+set(CONF_INCLUDE_DIRS "${CMAKE_INSTALL_PREFIX}/include")
```
This will point to the wrong include dirs if during configuration the user change the GNUInstallDirs's variable `CMAKE_INSTALL_INCLUDEDIR` . Furthermore, the include directories are already included in the imported target, so there are more maintainable ways of exporting this info ensuring that the package remains relocatable.

traversaro commented on cmake/abb_libegmConfig.cmake.in:
```diff
@@ -0,0 +1,16 @@
+# It defines the following variables
+#  abb_libegm_INCLUDE_DIRS - include directories for FooBar
+#  abb_libegm_LIBRARIES    - libraries to link against
+#  abb_libegm_EXECUTABLE   - the bar executable
```
I do not think `abb_libegm_EXECUTABLE` is defined anywhere.

jontje replied:
True. It's not defined and the comment can be removed. This comes from the ROS2 template [file](https://github.com/ros2/ros2cli/blob/master/ros2pkg/ros2pkg/resource/cmake/Config.cmake.in.em).

traversaro commented on cmake/abb_libegmConfig.cmake.in:
```diff
@@ -0,0 +1,16 @@
+#  abb_libegm_LIBRARIES    - libraries to link against
+#  abb_libegm_EXECUTABLE   - the bar executable
+
+set(abb_libegm_INCLUDE_DIRS "@CONF_INCLUDE_DIRS@")
```
If you are already exporting properly configured CMAke exported targets, there is little use in having an explicit variable `${abb_libegm_INCLUDE_DIRS}`, as any use of this variable in downstream package will make them non-relocatable. However, if you want to include it, you can extract it directly from the `abb_libegm` imported target: 
~~~
get_target_property(abb_libegm_INCLUDE_DIRS  abb_libegm INTERFACE_INCLUDE_DIRECTORIES)
~~~
clearly this line should go after the inclusion of  the `export_abb_libegm.cmake` file.

traversaro commented on cmake/abb_libegmConfig.cmake.in:
```diff
@@ -0,0 +1,16 @@
+set(abb_libegm_INCLUDE_DIRS "@CONF_INCLUDE_DIRS@")
+
+# Our library dependencies (contains definitions for IMPORTED targets)
+include("${abb_libegm_DIR}/export_abb_libegm.cmake")
```
The classical trick is to use `${CMAKE_CURRENT_LIST_DIR}` to import file that are in the same directory,to avoid assuming that a variable called `abb_libegm_DIR` is present.

traversaro replied:
See https://cmake.org/cmake/help/v3.3/manual/cmake-packages.7.html#creating-a-package-configuration-file .

jontje replied:
I see, thanks for the pointer!

gavanderhoorn commented:
A high-level comment: we'll have to decide which versions of OS we'd like to support here: Ubuntu Xenial (still supported till April 2021) comes with CMake `3.5.1` by default.

While CMake can be upgraded -- even using a Kitware supported apt repository -- this doesn't solve the problems of dependencies not necessarily being up-to-speed with modern CMake.

This PR in its current form would already exclude use of Xenial, due to the `cmake_minimum_required(VERSION 3.6)`.

gavanderhoorn commented:
@jontje: I've just added 1576b74 which adds a build for Dashing.

traversaro commented:
>  A high-level comment: we'll have to decide which versions of OS we'd like to support here: Ubuntu Xenial (still supported till April 2021) comes with CMake 3.5.1 by default.
> This PR in its current form would already exclude use of Xenial, due to the cmake_minimum_required(VERSION 3.6).

The fact that Xenial was already excluded is the reason why I suggest requiring CMake 3.9 . 
If we want instead to keep compatibility with Xenial and CMake 3.5, probably it would be too much effort to be prepare a relocatable package.

> While CMake can be upgraded -- even using a Kitware supported apt repository -- this doesn't solve the problems of dependencies not necessarily being up-to-speed with modern CMake.

If this was related to the Protobuf or Boost, what needs to be recent to have imported targets is not the Protobuf or Boost version, but the `FindBoost.cmake` and `FindProtobuf.cmake` shipped with CMake, and in their case the CMake imported targets are defined there and not in any Protobuf/Boost file.

jontje commented:
> A high-level comment: we'll have to decide which versions of OS we'd like to support here: Ubuntu Xenial (still supported till April 2021) comes with CMake `3.5.1` by default.

I was not considering that. The reason I put `3.6` is that I thought FindProtobuf set variables as `Protobuf_` instead of `PROTOBUF_`, according to [3.5](https://cmake.org/cmake/help/v3.5/module/FindProtobuf.html) and [3.6](https://cmake.org/cmake/help/v3.6/module/FindProtobuf.html) respectively.

But I tested with cmake `3.10.2` just now, and then it seems that both `Protobuf_` and `PROTOBUF_` variables are populated. Is this usually the case between cmake versions? I.e. for backwards-compatibility?

Regardless, then I think it would be good to still support Ubuntu Xenial from the get-go.

traversaro commented:
> Regardless, then I think it would be good to still support Ubuntu Xenial from the get-go.

Ack, then feel free to drop all the comments about non-relocatable packages, as the complexity of  doing that with CMake <= 3.9 is probably not worth the gain.

traversaro commented on include/abb_libegm/visibility_control.h:
```diff
@@ -0,0 +1,35 @@
+  #define ABB_LIBEGM_PUBLIC_TYPE
+#endif
+
+#endif  // ABB_LIBEGM__VISIBILITY_CONTROL_H_
```
I guess also this comes from the ros2 template, but unfortunately this template is based on the wrong section of the Wiki page https://gcc.gnu.org/wiki/Visibility . This is based on the section "How to use the new C++ visibility support" that is meant to be a quick overview of the functionality, while the template to be used is the one under "Step-by-step guide" . The difference are not enormous, but unfortunatly this version will probably create problems when compiling as static library in Windows, see https://github.com/ament/ament_cmake/issues/201#issuecomment-543898151 . Again, I do not care too much about building as a static library per se, but supporting that case is useful to easily submit a port of a C++ library to https://github.com/microsoft/vcpkg , that is a great way to manage C++ libraries on Windows.

jontje replied:
Yes, it comes from the ROS2 template and I will make an update to address this.

Also, thanks for all the insights! They are much appreciated!

traversaro replied:
> Yes, it comes from the ROS2 template and I will make an update to address this.

I imagined, I will try in the future to report this problems also there!

gavanderhoorn commented:
Interestingly the [ROS 2 build](https://travis-ci.com/ros-industrial/abb_libegm/jobs/256549698) succeeded. The [ROS 1 build](https://travis-ci.com/ros-industrial/abb_libegm/jobs/256549697) didn't.

I haven't looked into it yet.

jontje commented:
> Interestingly the [ROS 2 build](https://travis-ci.com/ros-industrial/abb_libegm/jobs/256549698) succeeded. The [ROS 1 build](https://travis-ci.com/ros-industrial/abb_libegm/jobs/256549697) didn't.
> 
> I haven't looked into it yet.

As far as I could see, then it is the cmake version `3.6` requirement that made the ROS 1 build fail, due to ROS Kinetic distribution.

gavanderhoorn commented:
> > Interestingly the [ROS 2 build](https://travis-ci.com/ros-industrial/abb_libegm/jobs/256549698) succeeded. The [ROS 1 build](https://travis-ci.com/ros-industrial/abb_libegm/jobs/256549697) didn't.
> > I haven't looked into it yet.
> 
> As far as I could see, then it is the cmake version `3.6` requirement that made the ROS 1 build fail, due to ROS Kinetic distribution.

yes, looks like it.

jontje commented:
@traversaro and @gavanderhoorn, I think I have addressed most of your comments (except the ones related to relocation).

Thanks a lot for your valuable remarks!

traversaro commented on CMakeLists.txt:
```diff
@@ -1,52 +1,64 @@
 #############################
-find_package(Protobuf REQUIRED)
+if(WIN32)
+  # The protobuf compiler appears to be default installed at ${_CMAKE_INSTALL_DIR}/tools/protobuf,
```
It may be worth to mention in the comment "appears to be default installed" refers to Chocolatey-based ROS/ROS2 installs.

traversaro commented on CMakeLists.txt:
```diff
@@ -55,30 +67,93 @@ set(SRC_FILES
+  # which is appropriate when building the dll but not consuming it.
+  target_compile_definitions(${PROJECT_NAME} PRIVATE "ABB_LIBEGM_BUILDING_SHARED_LIBRARY")
+else()
+  target_compile_definitions(${PROJECT_NAME} PRIVATE "ABB_LIBEGM_STATIC_LIBRARY")
```
```suggestion
  target_compile_definitions(${PROJECT_NAME} PUBLIC "ABB_LIBEGM_STATIC_LIBRARY")
```

`ABB_LIBEGM_STATIC_LIBRARY` needs to be defined also for the compilation unit that consume the library, otherwise static structure will be decorated with spurious `__declspec(dllimport)` .

traversaro commented on cmake/abb_libegmConfig.cmake.in:
```diff
@@ -0,0 +1,16 @@
+include("${CMAKE_CURRENT_LIST_DIR}/abb_libegmTargets.cmake")
+
+# Extras for our library (exports the dependency on pthread)
+include("${CMAKE_CURRENT_LIST_DIR}/abb_libegmExtras.cmake")
```
I just realized that `list(APPEND CMAKE_CXX_FLAGS -pthread)` contained in this file will probably break any downstream user of the package on Windows/MSVC that calls `find_package(abb_libegm REQUIRED)` , as `-pthread` is not a valid C++ flag in Windows/MSVC (cc @AlbertoRemusIIT) . I guess the CI is not testing this, so we did not realize that. 

Probably the CMake idiomatic way of dealing with this is to add `find_package(Threads REQUIRED)` and a  `target_link_libraries(${PROJECT_NAME} PUBLIC Threads::Threads)`, and then add:
~~~
include(CMakeFindDependencyMacro)
find_dependency(Threads REQUIRED)
~~~

here in the `abb_libegmConfig.cmake.in` before the inclusion of the `abb_libegmTargets.cmake` file.

traversaro commented on CMakeLists.txt:
```diff
@@ -55,30 +67,93 @@ set(SRC_FILES
+  ${Boost_LIBRARIES}
+  ${PROTOBUF_LIBRARIES}
+)
+
```
As you added the NAMESPACE to the generated imported target of CMake, it may be convenient to also add add a:
~~~
add_library(${PROJECT_NAME}::${PROJECT_NAME} ALIAS ${PROJECT_NAME})
~~~

This simplifies the use of the library with the CMake's FetchContent module, see https://github.com/open62541/open62541/issues/2436 and https://github.com/robotology/how-to-export-cpp-library/blob/1261ae57bc7c0422f0ab8b456df6f206f4935368/src/LibTemplateCMake/CMakeLists.txt#L62  .

jontje commented:
@traversaro thanks for your latest comments, I think I have considered them all.

The library builds fine and I have run RobotStudio simulations for both Ubuntu and Windows applications, so I will mark this PR ready for review.

traversaro commented:
> @traversaro thanks for your latest comments, I think I have considered them all.

Yes, thanks a lot for the effort!

traversaro commented:
Just to understand, for Windows you used protobuf installed from the ROS's Chocolatey distribution, right?

jontje commented:
> Just to understand, for Windows you used protobuf installed from the ROS's Chocolatey distribution, right?

Yes, that is right.

traversaro commented:
Just to let you know, we are having problems on Windows with Protobuf 3.10 (ROS repos have 3.6), but this probably due to an issue in Protobuf, not due to something wrong in `abb_libegm` ,
cc @AlbertoRemusIIT .

gavanderhoorn commented on CMakeLists.txt:
```diff
@@ -55,30 +73,87 @@ set(SRC_FILES
+add_library(${PROJECT_NAME}::${PROJECT_NAME} ALIAS ${PROJECT_NAME})
+target_include_directories(${PROJECT_NAME} PUBLIC
+  "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include;${CMAKE_CURRENT_BINARY_DIR}>"
+  $<INSTALL_INTERFACE:include>
```
```suggestion
  $<INSTALL_INTERFACE:$<INSTALL_PREFIX>/include>
```

gavanderhoorn commented on CMakeLists.txt:
```diff
@@ -55,30 +73,87 @@ set(SRC_FILES
 add_library(${PROJECT_NAME} ${SRC_FILES})
-target_include_directories(${PROJECT_NAME} PUBLIC ${CMAKE_CURRENT_BINARY_DIR})
-target_link_libraries(${PROJECT_NAME} ${Boost_LIBRARIES} ${PROTOBUF_LIBRARIES} ${catkin_LIBRARIES})
+add_library(${PROJECT_NAME}::${PROJECT_NAME} ALIAS ${PROJECT_NAME})
```
Suggestion: let's be a little more creative with the namespace and exported targets:

```suggestion
add_library(ABB::libegm ALIAS ${PROJECT_NAME})
```

gavanderhoorn replied:
We'd need to something like this as well to do this though:

```cmake
set_target_properties(${PROJECT_NAME} PROPERTIES EXPORT_NAME libegm)
```

traversaro replied:
I do not feel strongly about this, but most CMake package that I know and almost all `Find***.cmake` scripts in CMake, have the namespace that corresponds to the package name, i.e. if you have `find_package(<package>)`, then your imported targets are `<package>::lib1`, `<package>::lib2` . I don't think this is documented or recommended in any CMake docs, but I feel this would go against the  " principle of least astonishment" , https://en.wikipedia.org/wiki/Principle_of_least_astonishment .

jontje replied:
I like @traversaro's argument, but @gavanderhoorn's approach would "harmonize" better with e.g. [abb_librws](https://github.com/ros-industrial/abb_librws).

So, I am a little bit split about which to use.

gavanderhoorn replied:
@traversaro: least astonishment is certainly something to keep in mind. However, I always have to lookup what names a `Find*.cmake` defines, as even for 'old style' CMake, there are only conventions and quite a few libraries (even those for which find scripts are distributed with the 'official' CMake distribution) violate all of them (lower case, all upper case, snake-case, with or without dashes, etc, etc).

Proper documentation should be added in all cases I believe.

Capitalising the namespace was something I saw in [pabloariasal/modern-cmake-sample](https://github.com/pabloariasal/modern-cmake-sample/blob/cfda7b51ff915409fd61d4f2e75b63032bedb182/libjsonutils/CMakeLists.txt#L20) (from [this article](https://pabloariasal.github.io/2018/02/19/its-time-to-do-cmake-right)). Now that is `n==1` of course, so hardly convincing from that perspective, but I must say I like it.

Another example of where this is used is Boost. The main namespace is `Boost`, so we get `Boost::system`. Not `Boost_system::Boost_system`.

Having written that: it's just a name. If @jontje prefers least-astonishment, let's go with `abb_libegm::abb_libegm`.

traversaro replied:
> Capitalising the namespace was something I saw in [pabloariasal/modern-cmake-sample](https://github.com/pabloariasal/modern-cmake-sample/blob/cfda7b51ff915409fd61d4f2e75b63032bedb182/libjsonutils/CMakeLists.txt#L20) (from [this article](https://pabloariasal.github.io/2018/02/19/its-time-to-do-cmake-right)). Now that is `n==1` of course, so hardly convincing from that perspective, but I must say I like it.
> 
> Another example of where this is used is Boost. The main namespace is `Boost`, so we get `Boost::system`. Not `Boost_system::Boost_system`.

I am probably missing something, but as far as I can see, both example respect the convention that a package found by `find_package(<package>)` always exports CMake targets in the form of `<package>::<libname>` , in particular: 

| `<package>` |             `<libname>`            |  `<package>::<libname>` | 
|:-------------:|:-----------------:|:-------------------:|
| JSONUtils    | jsonutils            |  JSONUtils::jsonutils  |
| Boost          |  system             | Boost::system |

traversaro replied:
> Having written that: it's just a name. If @jontje prefers least-astonishment, let's go with `abb_libegm::abb_libegm`.

Same for me, bikeshedding is fun but I would feel fine with either decision.

jontje replied:
Let's go with `abb_libegm::abb_libegm`.

gavanderhoorn commented on CMakeLists.txt:
```diff
@@ -55,30 +73,87 @@ set(SRC_FILES
+    "${PROJECT_BINARY_DIR}/${CMAKE_FILES_DIRECTORY}/${PROJECT_NAME}Config.cmake"
+    "${PROJECT_BINARY_DIR}/${PROJECT_NAME}ConfigVersion.cmake"
+  DESTINATION "share/${PROJECT_NAME}/cmake"
+  COMPONENT dev
```
Do we really need this? Or is this a cargo-cult copy-pasta?

jontje replied:
It's not needed, I guess, it originates from the ROS2 cmake template.

gavanderhoorn commented on CMakeLists.txt:
```diff
@@ -55,30 +73,87 @@ set(SRC_FILES
+  EXPORT export_${PROJECT_NAME}
+  DESTINATION "share/${PROJECT_NAME}/cmake"
+  FILE ${PROJECT_NAME}Targets.cmake
+  NAMESPACE ${PROJECT_NAME}::
```
See my earlier comment about the namespace:

```suggestion
  NAMESPACE ABB::
```

gavanderhoorn commented on CMakeLists.txt:
```diff
@@ -55,30 +73,87 @@ set(SRC_FILES
+  FILES
+    "${PROJECT_BINARY_DIR}/${CMAKE_FILES_DIRECTORY}/${PROJECT_NAME}Config.cmake"
+    "${PROJECT_BINARY_DIR}/${PROJECT_NAME}ConfigVersion.cmake"
+  DESTINATION "share/${PROJECT_NAME}/cmake"
```
Should this not be:

```suggestion
  DESTINATION ${CMAKE_INSTALL_DATAROOTDIR}/${PROJECT_NAME}
```

?

jontje replied:
Less hard-coded the better, and it also originates from the ROS2 cmake template.

gavanderhoorn commented on CMakeLists.txt:
```diff
@@ -55,30 +73,87 @@ set(SRC_FILES
+
+install(
+  EXPORT export_${PROJECT_NAME}
+  DESTINATION "share/${PROJECT_NAME}/cmake"
```
Same here:
```suggestion
  DESTINATION ${CMAKE_INSTALL_DATAROOTDIR}/${PROJECT_NAME}
```

What is the rationale for hard-coding `share` here?

jontje replied:
Again, from the ROS2 cmake template.

gavanderhoorn commented on CMakeLists.txt:
```diff
@@ -55,30 +73,87 @@ set(SRC_FILES
+  DESTINATION "share/${PROJECT_NAME}/cmake"
+  FILE ${PROJECT_NAME}Targets.cmake
+  NAMESPACE ${PROJECT_NAME}::
+  COMPONENT dev
```
Needed?

jontje replied:
Again, from the ROS2 cmake template. I will remove it.

gavanderhoorn commented on CMakeLists.txt:
```diff
@@ -1,52 +1,70 @@
+
+# Work around Protobuf exporting 'lpthread' as a library: we let the
+# previous find_package(...) determine the system's thread library.
+list(REMOVE_ITEM PROTOBUF_LIBRARIES "-lpthread")
```
Have we verified that Protobuf is happy with this in all cases?

jontje replied:
I am not sure how to verify for all cases.

gavanderhoorn commented on cmake/abb_libegmConfig.cmake.in:
```diff
@@ -0,0 +1,18 @@
+find_dependency(Threads REQUIRED)
+
+# Our library dependencies (contains definitions for IMPORTED targets)
+include("${CMAKE_CURRENT_LIST_DIR}/abb_libegmTargets.cmake")
```
I don't believe we can hard-code this, as the name and path to this file are generated from values of CMake variables:

```cmake
  FILE "${PROJECT_BINARY_DIR}/${PROJECT_NAME}Targets.cmake"
```

We should probably use at least `@PROJECT_NAME@` here.

gavanderhoorn commented on cmake/abb_libegmConfig.cmake.in:
```diff
@@ -0,0 +1,18 @@
+get_target_property(abb_libegm_INCLUDE_DIRS abb_libegm::abb_libegm INTERFACE_INCLUDE_DIRECTORIES)
+
+# These are IMPORTED targets created by abb_libegmTargets.cmake
+set(abb_libegm_LIBRARIES abb_libegm::abb_libegm)
```
Can you explain why we do this?

Catkin does not need it: it knows how to deal with modern-cmake exported targets.

Other users should use at least CMake 3.5, so should also be able to use `target_link_libraries(..)`.

jontje replied:
The `abb_libegm_LIBRARIES` seems to be needed for using the `ament_target_dependencies` macro, to add dependencies to a ROS2 package's executable using the library.

If I remove the variable, then the build fails, unless I instead use the `target_link_libraries` macro.

But the `abb_libegm_INCLUDE_DIRS` can be removed.

gavanderhoorn replied:
I haven't checked, but is `ament_target_dependencies(..)` supposed to know how to deal with plain CMake packages?

If it is, then it should be fixed.

If it isn't, then I would suggest we don't export these old style CMake variables.

jontje replied:
> I haven't checked, but is `ament_target_dependencies(..)` supposed to know how to deal with plain CMake packages?

The example in `ament_cmake`'s user documentation for [adding deppendicies](https://index.ros.org/doc/ros2/Tutorials/Ament-CMake-Documentation/#adding-dependencies) is using `ament_target_dependencies` with `Eigen3`, which is a plain CMake package (as far as I know).

gavanderhoorn commented on package.xml:
```diff
@@ -1,22 +1,20 @@
   <version>1.0.0</version>
-  <description>The abb_libegm package for easing the use of ABB Externally Guided Motion (EGM) product</description>
-
+  <description>Package for easing the use of ABB's Externally Guided Motion (EGM) product</description>
```
```suggestion
  <description>A C++ library for interfacing with ABB robot controllers supporting Externally Guided Motion (689-1)</description>
```

gavanderhoorn commented on package.xml:
```diff
@@ -1,22 +1,20 @@
-  <author email="jon.tjerngren@se.abb.com">Jon Tjerngren</author>
-
   <url type="repository">https://github.com/ros-industrial/abb_libegm</url>
+  <author email="jon.tjerngren@se.abb.com">Jon Tjerngren</author>
```
The `maintainer` already has your email address. We typicaly only list it there. If you ever stop maintaining, it's convenient to just have to remove the `maintainer` tag, leave the `author` one and everything is immediately up-to-date (as opposed to having to edit/remove the email address in two places).

gavanderhoorn commented on package.xml:
```diff
@@ -1,22 +1,20 @@
-  <build_depend>boost</build_depend>
-  <build_depend>protobuf-dev</build_depend>
+  <depend>boost</depend>
+  <depend>protobuf-dev</depend>
```
If I interpret [the rosdep entries for protobuf](https://github.com/ros/rosdistro/blob/30a3b11485bc5948f08e0f93d4f17ff199e2b571/rosdep/base.yaml#L4607-L4645) correctly, I believe `protobuf-dev` would only be a `build_depend` and `build_export_depend`, and `protobuf` would still be the `exec_depend`.

The former resolves to the development package (providing the headers and compiler), while the latter contains just the library.

Headers are only needed for building things. The library is always needed.

jontje replied:
I took a peek at how Google Cartographer's [package.xml](https://github.com/googlecartographer/cartographer/blob/master/package.xml#L50) looked like for a newer package format and assumed it was the "appropriate" way.

But I think your way is more reasonable.

gavanderhoorn replied:
Well .. it *is* Google, but I'm not sure that is correct.

As I wrote: *if* I interpret things correctly.

gavanderhoorn commented on CMakeLists.txt:
```diff
@@ -1,52 +1,70 @@
+include(GNUInstallDirs)
+
+if(WIN32)
+  set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
```
Should we use `GenerateExportHeader` instead of a custom visibility header? See also [Create dlls on Windows without declspec() using new CMake export all feature](https://blog.kitware.com/create-dlls-on-windows-without-declspec-using-new-cmake-export-all-feature/).

jontje replied:
Great idea! I will add it.

gavanderhoorn commented on CMakeLists.txt:
```diff
@@ -55,30 +73,87 @@ set(SRC_FILES
+target_include_directories(${PROJECT_NAME} PUBLIC
+  "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include;${CMAKE_CURRENT_BINARY_DIR}>"
+  $<INSTALL_INTERFACE:include>
+  PUBLIC ${Boost_INCLUDE_DIRS}
```
The `FindBoost.cmake` from CMake `3.5` actually seems to export `Boost::` namespaced exported targets.

We should be able to just `target_link_libraries(.. Boost::system` etc)` instead of using `Boost_LIBRARIES` and `Boost_INCLUDE_DIRS`.

gavanderhoorn commented on CMakeLists.txt:
```diff
@@ -55,30 +73,87 @@ set(SRC_FILES
+  "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include;${CMAKE_CURRENT_BINARY_DIR}>"
+  $<INSTALL_INTERFACE:include>
+  PUBLIC ${Boost_INCLUDE_DIRS}
+  PUBLIC ${PROTOBUF_INCLUDE_DIRS}
```
You don't need to repeat `PUBLIC`.

gavanderhoorn commented on CMakeLists.txt:
```diff
@@ -55,30 +73,87 @@ set(SRC_FILES
+  PUBLIC ${PROTOBUF_INCLUDE_DIRS}
+)
+target_link_libraries(${PROJECT_NAME} PUBLIC
+  ${Boost_LIBRARIES}
```
See previous comment about `Boost::boost` and friends.

traversaro commented:
> 
> 
> Just to let you know, we are having problems on Windows with Protobuf 3.10 (ROS repos have 3.6), but this probably due to an issue in Protobuf, not due to something wrong in `abb_libegm` ,
> cc @AlbertoRemusIIT .

We investigate a bit more about this, and the issue is confirmed also when using the Chocolatey-installed protobuf used in ROS for Windows. If `abb_libegm` is compiled as a shared library, as soon as you try to link it to an executable, we get the following linker error: 
~~~
abb_libegm_test.obj : error LNK2019: unresolved external symbol "class abb::egm::wrapper::CartesianDefaultTypeInternal abb::egm::wrapper::_Cartesian_default_instance_" (?_Cartesian_default_instance_@wrapper@egm
@abb@@3VCartesianDefaultTypeInternal@123@A) referenced in function "void __cdecl ____C_A_T_C_H____T_E_S_T____11(void)" (?____C_A_T_C_H____T_E_S_T____11@@YAXXZ) [C:\ros-ws\idjl-software\build\src\libraries\idjl-
testsuite\idjl-testsuite.vcxproj]
abb_libegm_test.obj : error LNK2019: unresolved external symbol "class abb::egm::wrapper::CartesianPoseDefaultTypeInternal abb::egm::wrapper::_CartesianPose_default_instance_" (?_CartesianPose_default_instance_
@wrapper@egm@abb@@3VCartesianPoseDefaultTypeInternal@123@A) referenced in function "void __cdecl ____C_A_T_C_H____T_E_S_T____11(void)" (?____C_A_T_C_H____T_E_S_T____11@@YAXXZ) [C:\ros-ws\idjl-software\build\src
\libraries\idjl-testsuite\idjl-testsuite.vcxproj]
abb_libegm_test.obj : error LNK2019: unresolved external symbol "class abb::egm::wrapper::CartesianSpaceDefaultTypeInternal abb::egm::wrapper::_CartesianSpace_default_instance_" (?_CartesianSpace_default_instan
ce_@wrapper@egm@abb@@3VCartesianSpaceDefaultTypeInternal@123@A) referenced in function "void __cdecl ____C_A_T_C_H____T_E_S_T____11(void)" (?____C_A_T_C_H____T_E_S_T____11@@YAXXZ) [C:\ros-ws\idjl-software\build
\src\libraries\idjl-testsuite\idjl-testsuite.vcxproj]
abb_libegm_test.obj : error LNK2019: unresolved external symbol "class abb::egm::wrapper::EulerDefaultTypeInternal abb::egm::wrapper::_Euler_default_instance_" (?_Euler_default_instance_@wrapper@egm@abb@@3VEule
rDefaultTypeInternal@123@A) referenced in function "void __cdecl ____C_A_T_C_H____T_E_S_T____11(void)" (?____C_A_T_C_H____T_E_S_T____11@@YAXXZ) [C:\ros-ws\idjl-software\build\src\libraries\idjl-testsuite\idjl-t
estsuite.vcxproj]
abb_libegm_test.obj : error LNK2019: unresolved external symbol "class abb::egm::wrapper::FeedbackDefaultTypeInternal abb::egm::wrapper::_Feedback_default_instance_" (?_Feedback_default_instance_@wrapper@egm@ab
b@@3VFeedbackDefaultTypeInternal@123@A) referenced in function "void __cdecl ____C_A_T_C_H____T_E_S_T____11(void)" (?____C_A_T_C_H____T_E_S_T____11@@YAXXZ) [C:\ros-ws\idjl-software\build\src\libraries\idjl-test
suite\idjl-testsuite.vcxproj]
abb_libegm_test.obj : error LNK2019: unresolved external symbol "class abb::egm::wrapper::HeaderDefaultTypeInternal abb::egm::wrapper::_Header_default_instance_" (?_Header_default_instance_@wrapper@egm@abb@@3VH
eaderDefaultTypeInternal@123@A) referenced in function "void __cdecl ____C_A_T_C_H____T_E_S_T____11(void)" (?____C_A_T_C_H____T_E_S_T____11@@YAXXZ) [C:\ros-ws\idjl-software\build\src\libraries\idjl-testsuite\id
jl-testsuite.vcxproj]
abb_libegm_test.obj : error LNK2019: unresolved external symbol "class abb::egm::wrapper::RobotDefaultTypeInternal abb::egm::wrapper::_Robot_default_instance_" (?_Robot_default_instance_@wrapper@egm@abb@@3VRobo
tDefaultTypeInternal@123@A) referenced in function "void __cdecl ____C_A_T_C_H____T_E_S_T____11(void)" (?____C_A_T_C_H____T_E_S_T____11@@YAXXZ) [C:\ros-ws\idjl-software\build\src\libraries\idjl-testsuite\idjl-t
estsuite.vcxproj]
C:\ros-ws\idjl-software\build\bin\Release\idjl-testsuite.exe : fatal error LNK1120: 7 unresolved externals [C:\ros-ws\idjl-software\build\src\libraries\idjl-testsuite\idjl-testsuite.vcxproj]
~~~

As soon as `abb_libegm` is compiled as a static library, everything works fine. So, unless I am missing something, the library was never tested by linking to an executable when compiled as shared in Windows? 

Some investigation in the issue revealed that the problems is that the Protobuf generated  code include some static variables, that need to be exported manually with the visibility macro as they are  not automatically exported by the `WINDOWS_EXPORT_ALL_SYMBOLS` target property, as mentioned in https://blog.kitware.com/create-dlls-on-windows-without-declspec-using-new-cmake-export-all-feature/ . 

Fortunately, the protobuf generator of C++ files permits to declare a decorator to add to all the symbols to export via the `dllexport_decl` command line option, that is also exposed as a `EXPORT_MACRO` in the `protobuf_generate_cpp` CMake function ( https://cmake.org/cmake/help/v3.9/module/FindProtobuf.html ).

Unfortunately, if we add in the CMake build  system the following line: 
~~~
protobuf_generate_cpp(EgmProtoSources EgmProtoHeaders EXPORT_MACRO ABB_LIBEGM_PUBLIC ${EgmProtoFiles})
~~~
the compilation fails, because in the the `.pb.h` and in the `.pb.cc` files, there is no option to specify an additional include header, in our case `<abb_libegm/visibility_control.hh>` or `<visibility_control.hh>` (see https://github.com/protocolbuffers/protobuf/issues/4198). I checked a bit how other projects deal with this, and Gazebo and Ignition Msgs actually implement their own C++ protobuf-based code generator to include the visibility headers (see https://bitbucket.org/osrf/gazebo/src/804410860234af97c5e309896dc007e8cde04ba8/gazebo/msgs/generator/GazeboGenerator.cc#lines-72 and https://bitbucket.org/ignitionrobotics/ign-msgs/src/18640ef6c31777e25953ae44b6ad10ed68c4993e/src/Generator.cc#lines-106), but that seems definitely an overkill for this project. There may be other solutions, such as do a post-processing of  the protobuf generated headers to add the include headers via CMake, but I feel they are out of the scope of this PR, and for the time being the easiest solution is to force the library to be compiled as static in Windows.

jontje commented:
> As soon as `abb_libegm` is compiled as a static library, everything works fine. So, unless I am missing something, the library was never tested by linking to an executable when compiled as shared in Windows?

I have linked the libary as shared to an executable in Windows and got it to work. But I have tested a bit more and if I use the `mutable_<field>` accessors, then it works but not with the `<field>` version. Then I get similar errors as you posted.

E.g.:
```
abb::egm::wrapper::Output output;
output.mutable_robot()->mutable_joints()->mutable_position()->values_size(); // Works
output.robot().joints().position().values_size(); // Doesn't work
```

> Fortunately, the protobuf generator of C++ files permits to declare a decorator to add to all the symbols to export via the `dllexport_decl` command line option, that is also exposed as a `EXPORT_MACRO` in the `protobuf_generate_cpp` CMake function ( https://cmake.org/cmake/help/v3.9/module/FindProtobuf.html ).

Is this only for `CMake 3.9` and newer? If I looked at older documentation that option doesn't seem to exist. Would it be possible to add some if-case for Windows which uses a macro that expands to either `__declspec(dllexport)` or `__declspec(dllimport)`? Does it even make sense?

> There may be other solutions, such as do a post-processing of the protobuf generated headers to add the include headers via CMake, but I feel they are out of the scope of this PR, and for the time being the easiest solution is to force the library to be compiled as static in Windows.

Yes, that might be the easiest solution for now. So something like this? To make static as default on windows.

```
if(NOT DEFINED BUILD_SHARED_LIBS)
  if(WIN32)
    option(BUILD_SHARED_LIBS "Build dynamically-linked binaries" OFF)
  else()
    option(BUILD_SHARED_LIBS "Build dynamically-linked binaries" ON)
  endif()
endif()
```

traversaro commented:
> I have linked the libary as shared to an executable in Windows and got it to work. But I have tested a bit more and if I use the `mutable_<field>` accessors, then it works but not with the `<field>` version. Then I get similar errors as you posted.

That is interesting, and in that executable you included the `.pb.h` headers of the generated protobuf messages? However, this may suggest to it could make sense to add a small smoke compilation test to the repo only when `BUILD_TESTING` is defined, to make sure that we don't have regression of this kind (the test can also just be used to verify that the linking works fine, and not do anything when executed).

traversaro commented:
> Yes, that might be the easiest solution for now. So something like this? To make static as default on windows.
> 
> ```
> if(NOT DEFINED BUILD_SHARED_LIBS)
>   if(WIN32)
>     option(BUILD_SHARED_LIBS "Build dynamically-linked binaries" OFF)
>   else()
>     option(BUILD_SHARED_LIBS "Build dynamically-linked binaries" ON)
>   endif()
> endif()
> ```

As `BUILD_SHARED_LIBS` is broken on Windows, I would probably just avoid to set it as an option, so simply: 
 
```cmake
 if(NOT DEFINED BUILD_SHARED_LIBS)
   if(WIN32)
       # Add a reference to an issue on the problem with protobuf
     set(BUILD_SHARED_LIBS  OFF)
   else()
     set(BUILD_SHARED_LIBS "Build dynamically-linked binaries" ON)
   endif()
endif()
```

traversaro commented:
> > Fortunately, the protobuf generator of C++ files permits to declare a decorator to add to all the symbols to export via the `dllexport_decl` command line option, that is also exposed as a `EXPORT_MACRO` in the `protobuf_generate_cpp` CMake function ( https://cmake.org/cmake/help/v3.9/module/FindProtobuf.html ).
> 
> Is this only for `CMake 3.9` and newer? If I looked at older documentation that option doesn't seem to exist. Would it be possible to add some if-case for Windows which uses a macro that expands to either `__declspec(dllexport)` or `__declspec(dllimport)`? Does it even make sense?

Yes, I think it make sense to have a check on the CMake version that is greater then 3.9 to add that option. Furthremore, it is also possible to force that the cmake minimum required version is 3.9 on Windows, as typically Windows user do not have any requirement on being compatible with older CMake versions.

traversaro commented:
Based on the answer in https://stackoverflow.com/questions/17980467/protocol-buffers-generate-non-inline-accessors , I was able to fix the linking problem for shared library in MSVC by adding the following CMake code:
~~~cmake
if(MSVC)
  cmake_minimum_required(VERSION 3.9)
else()
  cmake_minimum_required(VERSION 3.5)
endif()

# ... 

if(MSVC)
  protobuf_generate_cpp(EgmProtoSources EgmProtoHeaders EXPORT_MACRO ABB_LIBEGM_PUBLIC ${EgmProtoFiles})
else()
  protobuf_generate_cpp(EgmProtoSources EgmProtoHeaders ${EgmProtoFiles})
endif()

# ...

if(MSVC)
  target_compile_options(${PROJECT_NAME} PUBLIC "/FIabb_libegm/visibility_control.h")
endif()
~~~

gavanderhoorn commented:
This is not directed at anyone here, but really, how ugly and convoluted can we make a build script?

And then people complain that Catkin / Ament are unnecessary / convoluted / too complex.

I'm actually happy that we have a set of CMake functions/macros that takes care of some of the boilerplate stuff when exporting and importing dependencies.

traversaro commented:
> 
> 
> This is not directed at anyone here, but really, how ugly and convoluted can we make a build script?
> 
> And then people complain that Catkin / Ament are unnecessary / convoluted / too complex.
> 
> I'm actually happy that we have a set of CMake functions/macros that takes care of some of the boilerplate stuff when exporting and importing dependencies.

Related to that, there are some related issues in upstream CMake on how to simplify the process of exporting CMake packages, that if implemented would reduce drastically the amount of boilerplate code: https://gitlab.kitware.com/cmake/cmake/issues/18634 .

jontje commented:
> However, this may suggest to it could make sense to add a small smoke compilation test to the repo only when `BUILD_TESTING` is defined, to make sure that we don't have regression of this kind (the test can also just be used to verify that the linking works fine, and not do anything when executed).

That's a good idea, but maybe do it in another PR?

jontje commented:
> This is not directed at anyone here, but really, how ugly and convoluted can we make a build script?

Any suggestion for solving the issue?

Maybe require `CMake 3.9` anyway? Is it "bad practice" to hide away code in a cmake module? If not, perhaps that would make the `CMakeLists` file clearer.

To me it seems a bit tricky to get the library to build properly, without some unfortunate special cases. Regardless, I really appreciate all the time you guys (@gavanderhoorn and @traversaro) have spent on this PR so far.

traversaro commented:
> That's a good idea, but maybe do it in another PR?

Yes totally, also for the Windows+Shared I think it make sense to do that in another PR.

traversaro commented on CMakeLists.txt:
```diff
@@ -1,52 +1,70 @@
+#########################
+## Boost C++ Libraries ##
+#########################
+find_package(Boost REQUIRED COMPONENTS regex system thread)
```
I noticed that the `regex` component is not used at all, and it was not present in the previous build system. 
```suggestion
find_package(Boost REQUIRED COMPONENTS system thread)
```

jontje replied:
I can't build the libary on Windows (at least with ROS and `catkin_make_isolated`) if I don't specify `regex`. I don't really know why and I haven't had time to look into it.

I get this linking error:

```
LINK Pass 1: command "C:\PROGRA~2\MICROS~3\2017\PROFES~1\VC\Tools\MSVC\1416~1.270\bin\Hostx64\x64\link.exe /nologo @CMakeFiles\abb_libegm.dir\objects1.rsp /out:abb_libegm.dll /implib:abb_libegm.lib /pdb:C:\win_ros\ros\win_catkin_ws\build_isolated\abb_libegm\devel\abb_libegm.pdb /dll /version:0.0 /machine:x64 /debug /INCREMENTAL /DEF:CMakeFiles\abb_libegm.dir\exports.def C:\opt\rosdeps\x64\lib\boost_system-vc141-mt-x64-1_66.lib C:\opt\rosdeps\x64\lib\boost_thread-vc141-mt-x64-1_66.lib C:\opt\rosdeps\x64\lib\libprotobuf.lib C:\opt\rosdeps\x64\lib\boost_chrono-vc141-mt-x64-1_66.lib C:\opt\rosdeps\x64\lib\boost_system-vc141-mt-x64-1_66.lib C:\opt\rosdeps\x64\lib\boost_date_time-vc141-mt-x64-1_66.lib C:\opt\rosdeps\x64\lib\boost_atomic-vc141-mt-x64-1_66.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:CMakeFiles\abb_libegm.dir/intermediate.manifest CMakeFiles\abb_libegm.dir/manifest.res" failed (exit code 1104) with the following output:
LINK : fatal error LNK1104: cannot open file 'boost_regex-vc141-mt-x64-1_66.lib'
```

traversaro replied:
Sorry, I had no idea, probably there is some non-declared internal boost dependency, in that case indeed leaving regex is indeed a nice workaround.

jontje commented:
I have created a new [branch](https://github.com/jontje/abb_libegm/tree/plain_cmake_draft) that I think would clean-up the CMake files, as well as added @traversaro's example for fixing the linking issue on Windows. I didn't add it to this PR yet, since I has been waiting for some comments from @gavanderhoorn, but I think this will have to wait until next year.

gavanderhoorn commented:
Getting back to this I think we can merge this as-is.

One comment: this will break peoples workspaces when they are currently using `abb_libegm` and building it with `catkin_make` (as plain `catkin` doesn't support plain cmake packages).

We should add a section to the readme explaining this is a pure CMake package and that ROS users should use `catkin_tools` to build it (or colcon).

gavanderhoorn commented:
I'll leave the merge to you @jontje.

Could you also update the readme?

We should also post about this breaking change on ROS Discourse. I can do that if you'd like. Let me know.

gavanderhoorn commented on CMakeLists.txt:
```diff
@@ -1,52 +1,70 @@
-list (REMOVE_ITEM PROTOBUF_LIBRARIES "-lpthread")
+if(WIN32)
+  # For a Chocolatey-based ROS/ROS2 installation, then it appears that the protobuf compiler is
+  # by default installed at ${_CMAKE_INSTALL_DIR}/tools/protobuf, and find_package() doesn't
```
Just noticed this: `_CMAKE_INSTALL_DIR`.

Is that actually a variable?

traversaro replied:
Yes, even if a non-documented/unsupported one (see https://gitlab.kitware.com/cmake/cmake/blob/v3.16.2/Modules/Platform/WindowsPaths.cmake#L49). It works fine only because in Chocolatey-based ROS/ROS2 installations CMake and protobuf are installed in the same install prefix.

gavanderhoorn replied:
>  even if a non-documented/unsupported one

hm ...

traversaro replied:
> 
> 
> > even if a non-documented/unsupported one
> 
> hm ...

Not a big fan of it either. : )

gavanderhoorn commented on CMakeLists.txt:
```diff
@@ -55,30 +73,89 @@ set(SRC_FILES
-        FILES_MATCHING PATTERN "*.h")
+install(
+  FILES
+    ${EgmProtoFiles}
```
Should these go to `CMAKE_INSTALL_DATAROOTDIR` instead?

traversaro replied:
I am not sure if  there is any policy/best practice for this, but for example the ignition-msgs package installs them in the same include directories of C++ include headers, see  https://packages.debian.org/sid/amd64/libignition-msgs-dev/filelist  . Perhaps @j-rivero knows about some policy regarding this.

jontje commented:
Thanks for the reviews @traversaro and @gavanderhoorn! 👍 

> I'll leave the merge to you @jontje.
> 
> Could you also update the readme?
> 
> We should also post about this breaking change on ROS Discourse. I can do that if you'd like. Let me know.

I will update the readme, and you are more than welcome to post on ROS Discourse this time @gavanderhoorn 😄

gavanderhoorn commented:
Done: [Conversion to plain CMake of abb_libegm (and abb_librws)](https://discourse.ros.org/t/conversion-to-plain-cmake-of-abb-libegm-and-abb-librws/12441).

gavanderhoorn commented:
You will update the readme?

jontje commented:
> You will update the readme?

Yes, I will do that.
